felix/dataplane/windows/endpoint_mgr.go

```diff
@@ -192,15 +191,7 @@ func (m *endpointManager) RefreshHnsEndpointCache(forceRefresh bool) error {
 		// Some CNI plugins do not clear endpoint properly when a pod has been torn down.
 		// In that case, it is possible Felix sees multiple endpoints with the same IP.
 		// We need to filter out inactive endpoints that do not attach to any container.
-		containers, err := m.hns.GetAttachedContainerIDs(&endpoint)
-		if err != nil {
-			log.WithFields(log.Fields{
-				"id":   endpoint.Id,
-				"name": endpoint.Name,
-			}).Warn("Failed to get attached containers")
-			continue
-		}
-		if len(containers) == 0 {
+		if len(endpoint.SharedContainers) == 0 {
```

Comment:
I am sorry I guess this is caught late. Do we need to make a `hnsCall("GET", "/endpoints/"+endpoint.Id, "", &attachInfo)` call before we check `SharedContainers`?

Reply:
Calling `IsAttached()` wouldn't work because it requires a specific container ID as an arg (so we'd need to somehow list all containers and then loop through them calling `IsAttached()` for each of them, per endpoint). But I think that's not necessary.

The endpoints are retrieved via an `hnsCall()` which happens inside `HNSListEndpointRequest()`, which in turn is called here: https://github.com/projectcalico/calico/blob/59b28041cacd4819dcea4224e08686482cabc882/felix/dataplane/windows/endpoint_mgr.go#L162

I couldn't find explicit documentation on `hnsCall()`, but I spun up a windows cluster and used the `hnsdiag` tool on powershell and verified that it does populate `SharedContainers`:
```
PS C:\C\e815cc9c19ec519569d4eeabe514771c61f1531e2c443299a420f27e8bcc0dd8\CalicoWindows> hnsdiag list endpoints -d
(...)
    "SharedContainers":[
        "1c809ef2e5d94f56600fe3c217e28838ba7641d95f2a3cd566126912ec1dee2c"
    ]
}
```

The `SharedContainers` must be ultimately being unmarshalled from the response here: https://github.com/projectcalico/hcsshim/blob/master/internal/hns/hnsfuncs.go#L43